src/main/java/net/neoforged/neoforge/common/CommonHooks.java

```diff
@@ -1341,10 +1340,30 @@ public static void validateComponent(@Nullable Object dataComponent) {
 
             if (overridesEqualsAndHashCode(clazz)) {
                 checkedComponentClasses.add(clazz);
-            } else {
-                throw new IllegalArgumentException("Data components must implement equals and hashCode. Keep in mind they must also be immutable. Problematic class: " + clazz);
+                return;
+            }
+
+            // By far the slowest check: Is this a registry object?
+            // If it is, we assume it must be usable as a singleton...
+            if (isPotentialRegistryObject(dataComponent)) {
+                checkedComponentClasses.add(clazz);
+                return;
+            }
+
+            throw new IllegalArgumentException("Data components must implement equals and hashCode. Keep in mind they must also be immutable. Problematic class: " + clazz);
+        }
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    private static boolean isPotentialRegistryObject(Object value) {
+        for (Registry registry : BuiltInRegistries.REGISTRY) {
+            if (registry.containsValue(value)) {
+                LOGGER.info("Whitelisting {} as component type, since {} is contained in registry {}",
```

Comment:
This is going to be spammy. Think how many times this will be checked in a modpack, at random times. I'd make it a debug log.

Reply:
Ah yes you're right... It's still low value I would say.